src/lua/uri.lua

```diff
@@ -220,6 +267,57 @@ local function values(...)
     return {...}
 end
 
+local char_to_hex = function(c)
+    return string.format("%%%02X", string.byte(c))
+end
+
+-- Encodes a string into its escaped hexadecimal representation.
+local function escape(buf, opts)
+    if type(buf) ~= "string" then
+        error("Usage: uri.escape(string, [opts])")
+    end
+    local options = opts or RFC3986
+    options.unreserved = options.unreserved or RFC3986.unreserved
+    options.plus = options.plus or RFC3986.plus
+
+    local pattern = unreserved(options.unreserved)
+    local unreserved = {}
+    for i = 0, 255 do
+        local ch = string.char(i)
+        unreserved[ch] = ch:match(pattern)
+    end
+
+    return (buf:gsub("(.)", function(ch)
+        if options.plus and ch == " " then
+            return "+"
+        elseif unreserved[ch] then
+            return ch
+        else
+            return char_to_hex(ch)
+        end
+    end))
+end
+
+local hex_to_char = function(x)
+    return string.char(tonumber(x, 16))
+end
+
+-- Decodes an escaped hexadecimal string into its binary representation.
+local function unescape(buf, opts)
+    if type(buf) ~= "string" then
+        error("Usage: uri.unescape(string, [opts])")
+    end
+    opts = opts or RFC3986
+    opts.plus = opts.plus or RFC3986.plus
+
+    local str = buf
+    if opts.plus == true then
+        str = str:gsub("+", " ")
+    end
```

Comment:
Handling of the `plus` option is a bit inconsistent with `uri.escape()`: `if plus then <...> end` vs `if plus == true then <...> end`.

I suggest to stick with `if plus then <...> end` for both functions. I'll explain the reasonsing below.

----

Let's consider a general case: we have a function `foo` that accepts a boolean option `bar`. There are the following variants of handling it:

1. Strict validation: `type(bar)` is `nil` or `boolean`. After the strict validation it doesn't matter, whether `== true` is used.
2. No type validation, check for `if bar then <...> end`.
3. No type validation, check for `if bar == true then <...> end`.

I think that 1 and 2 are okay, 3 is undesirable, inconsistent handling is bad.

Why 3 is undesirable? Let's show an example:

```lua
local function foo(opts)
    local opts = opts or {}
    local is_request_polite = opts.is_request_polite

    if is_request_polite == true then
        return 'ok'
    end

    return 'no'
end

local is_request_polite = request:find('please') -- number or nil
local res = process_request({is_request_polite = is_request_polite})
```

This code always give `'no'`.

We sometimes use non-boolean expressions as predicates. It is okay to push a user of the API to explicitly cast a value to boolean by issuing an error (variant 1). It is okay to accept a non-boolean value with the same meaning as it has in the `if` statement -- it is quite natural for a Lua developer. But it may be confusing and error prone that, say, `42` is interpreted as a false value.